src/lib.rs

```diff
@@ -0,0 +1,626 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+use async_trait::async_trait;
+use prost::Message;
+use std::collections::HashMap;
+use std::sync::{
+    atomic::{AtomicU64, Ordering},
+    Arc, Mutex,
+};
+use std::time::Duration;
+use uprotocol_sdk::{
+    rpc::{RpcClient, RpcClientResult, RpcMapperError, RpcServer},
+    transport::{datamodel::UTransport, validator::Validators},
+    uprotocol::{
+        Data, UAttributes, UCode, UEntity, UMessage, UMessageType, UPayload, UPayloadFormat,
+        UPriority, UStatus, UUri,
+    },
+    uri::{
+        serializer::{MicroUriSerializer, UriSerializer},
+        validator::UriValidator,
+    },
+};
+use zenoh::{
+    config::Config,
+    prelude::{r#async::*, Sample},
+    queryable::{Query, Queryable},
+    sample::AttachmentBuilder,
+    subscriber::Subscriber,
+};
+
+pub struct ZenohListener {}
+pub struct UPClientZenoh {
+    session: Arc<Session>,
+    subscriber_map: Arc<Mutex<HashMap<String, Subscriber<'static, ()>>>>,
+    queryable_map: Arc<Mutex<HashMap<String, Queryable<'static, ()>>>>,
+    query_map: Arc<Mutex<HashMap<String, Query>>>,
+    callback_counter: AtomicU64,
+}
+
+impl UPClientZenoh {
+    /// # Errors
+    /// Will return `Err` if unable to create Zenoh session
+    pub async fn new(config: Config) -> Result<UPClientZenoh, UStatus> {
+        let Ok(session) = zenoh::open(config).res().await else {
+            return Err(UStatus::fail_with_code(
+                UCode::Internal,
+                "Unable to open Zenoh session",
+            ));
+        };
+        Ok(UPClientZenoh {
+            session: Arc::new(session),
+            subscriber_map: Arc::new(Mutex::new(HashMap::new())),
+            queryable_map: Arc::new(Mutex::new(HashMap::new())),
+            query_map: Arc::new(Mutex::new(HashMap::new())),
+            callback_counter: AtomicU64::new(0),
+        })
+    }
+
+    fn to_zenoh_key_string(uri: &UUri) -> Result<String, UStatus> {
+        let micro_uuri = MicroUriSerializer::serialize(uri).map_err(|_| {
+            UStatus::fail_with_code(
+                UCode::InvalidArgument,
+                "Unable to serialize into micro format",
+            )
+        })?;
+        let micro_zenoh_key: String = micro_uuri
+            .iter()
+            .fold(String::new(), |s, c| s + &format!("{c:02x}"));
```

Comment:
I've been thinking about this more in the context of usage by a uStreamer in support of [the special mode I mention here](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/pull/2/files#r1463538303).

In that case, in order to make it easier to implement the `register_listener(all_with_UAuthority)` functionality, I think choosing an alternative way to arrange the Zenoh key in that case could be beneficial. IMHO we don't need to have the Zenoh key match the Micro URI, so long as we document how we're doing it.

The Micro URI form [as documented here](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc#42-micro-uris) has the UAuthority (if any) appended on the end.

However if I ask to `register_listener(all_with_UAuthority)`, implementing that via a Zenoh key could become challenging in `up-client-zenoh-rust`.

Perhaps in addition to placing `up/` in front, we could move the UAuthority in front like this: `up/[byte representation of UAuthority]/[byte representation of rest of UUri]`.

Someone calling `register_listener(all_with_UAuthority)` could have their Zenoh key then be: `up/*/**` to ensure they got only messages with a UAuthority attached.

WDYT?

Reply:
@PLeVasseur I don't quite get the idea why we need to separate UAuthourity in Zenoh key. Can't we just use `up/**`?

Reply:
Oh, because [not every Micro Uri must contain a UAuthority](https://github.com/eclipse-uprotocol/up-spec/blob/main/basics/uri.adoc#42-micro-uris):

![image](https://github.com/eclipse-uprotocol/up-client-zenoh-rust/assets/11622119/0bcc3fca-1cf8-4e64-ab46-217be3897feb)

And the order that the bytes are written would normally put the bytes for UAuthority at the end.

Reply:
I see, so you want to only listen to the Uuri with remote UAuthority and don't need to filter it in the `register_listener`. It seems OK to me. Maybe we can raise the discussion in tomorrow's meeting in case we miss anything.